gr-analog/lib/sig_source_impl.cc

```diff
@@ -64,32 +67,77 @@ sig_source_impl<T>::~sig_source_impl()
 }
 
 template <class T>
-void sig_source_impl<T>::set_frequency_msg(pmt::pmt_t msg)
+void sig_source_impl<T>::set_freq_msg(pmt::pmt_t msg)
 {
-    // Accepts either a number that is assumed to be the new
-    // frequency or a key:value pair message where the key must be
-    // "freq" and the value is the new frequency.
-
+    GR_LOG_INFO(this->d_logger,
+                "The `freq` port is deprecated and will be removed. Forwarding this "
+                "message to the `cmd` handler.");
     if (pmt::is_number(msg)) {
+        // if plain number, then interpret it as frequency
         set_frequency(pmt::to_double(msg));
+        return;
+    }
+    set_cmd_msg(msg);
+}
+
+template <class T>
+void sig_source_impl<T>::set_cmd_msg(pmt::pmt_t msg)
+{
+    static auto freq_key = pmt::intern("freq");
+    static auto ampl_key = pmt::intern("ampl");
+    static auto phase_key = pmt::intern("phase");
+    static auto offset_key = pmt::intern("offset");
+
+    // either a key:value pair or a dict
+    pmt::pmt_t list_of_items;
+    if (pmt::is_dict(msg)) {
+        list_of_items = pmt::dict_items(msg);
     } else if (pmt::is_pair(msg)) {
-        pmt::pmt_t key = pmt::car(msg);
-        pmt::pmt_t val = pmt::cdr(msg);
-        if (pmt::eq(key, pmt::intern("freq"))) {
+        list_of_items = pmt::list1(msg);
+    } else {
+        GR_LOG_WARN(this->d_logger,
+                    "malformed message: is neither plain number, nor dict or pair");
```

Comment:
Plain numbers are only accepted by the calling set_freq_msg function. Should this message include plain numbers?

Reply:
no, it shouldn't!